prboom2/src/MUSIC/libopenmptplayer.c

```diff
@@ -0,0 +1,184 @@
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
+ *  02111-1307, USA.
+ *
+ * DESCRIPTION:
+ *  Tracker music player based on libopenmpt.
+ *---------------------------------------------------------------------
+ */
+
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "musicplayer.h"
+
+#ifndef HAVE_LIBOPENMPT
+#include <stddef.h>
+
+static const char *mpt_name (void)
+{
+  return "libopenmpt tracker player (DISABLED)";
+}
+
+
+static int mpt_init (int samplerate)
+{
+  return 0;
+}
+
+const music_player_t mpt_player =
+{
+  mpt_name,
+  mpt_init,
+  NULL,
+  NULL,
+  NULL,
+  NULL,
+  NULL,
+  NULL,
+  NULL,
+  NULL,
+  NULL
+};
+
+#else // HAVE_LIBOPENMPT
+
+#include <libopenmpt/libopenmpt.h>
+#include <string.h>
+#include <stdlib.h>
+#include "lprintf.h"
+
+
+static int mpt_samplerate;
+static int mpt_paused;
+static float mpt_multiplier;
+static openmpt_module *mpt_cur_module;
+
+static const char *mpt_name (void)
+{
+  return "libopenmpt tracker player";
+}
+
+
+static int mpt_init (int samplerate)
+{
+  mpt_samplerate = samplerate;
+
+  return 1;
+}
+
+static void mpt_shutdown (void)
+{
+}
+
+static void mpt_setvolume (int v)
+{
+  mpt_multiplier = 16384.f / 15.f * v;
+}
+
+static const void* mpt_registersong (const void *data, unsigned len)
+{
+  openmpt_module *mod = NULL;
+
+  mod = openmpt_module_create_from_memory2(data, len, NULL, NULL, NULL, NULL, NULL, NULL, NULL);
+  if (!mod)
+    lprintf(LO_WARN, "mpt_registersong: failed to open song\n");
+
+  return mod;
+}
+
+static void mpt_unregistersong (const void *handle)
+{
+  #pragma GCC diagnostic push
+  #pragma GCC diagnostic ignored "-Wcast-qual"
+  openmpt_module_destroy((openmpt_module *) handle);
+  #pragma GCC diagnostic pop
+}
+
+static void mpt_play (const void *handle, int looping)
+{
+  int ret;
+
+  #pragma GCC diagnostic push
+  #pragma GCC diagnostic ignored "-Wcast-qual"
+  mpt_cur_module = (openmpt_module *) handle;
+  #pragma GCC diagnostic pop
+  ret = openmpt_module_set_repeat_count(mpt_cur_module, looping ? -1 : 0);
+
+  if (!ret)
+    mpt_cur_module = NULL;
+}
+
+static void mpt_stop (void)
+{
+  mpt_cur_module = NULL;
+}
+
+static void mpt_pause (void)
+{
+  mpt_paused = 1;
+}
+
+static void mpt_resume (void)
+{
+  mpt_paused = 0;
+}
+
+static void mpt_render (void *dest, unsigned nsamp)
+{
+  short *sdest = (short *)dest;
+
+  if (mpt_cur_module && !mpt_paused)
+  {
+    size_t nout, i;
+    float *buf = malloc(sizeof(float) * nsamp * 2);
+
+    nout = openmpt_module_read_interleaved_float_stereo(mpt_cur_module,
+                    mpt_samplerate, nsamp, buf);
+
+    for (i = 0; i < nout; i++)
+    {
+      sdest[i * 2 + 0] = (short) (buf[i * 2 + 0] * mpt_multiplier);
+      sdest[i * 2 + 1] = (short) (buf[i * 2 + 1] * mpt_multiplier);
```

Comment:
How about setting gain with [openmpt_module_set_render_param](https://lib.openmpt.org/doc/group__libopenmpt__c.html#gaecd0bb67bb3b720340bb12f5df86db2a) instead of using `mpt_multiplier`?